XMLParser/XMLParser.cpp

```diff
@@ -0,0 +1,361 @@
+		}
+		else if (line[tag_begin + one] != line.size() && line[tag_begin + one] == symbol->slash)
+		{
+			close_tag = line.substr(tag_begin + two, tag_end - (tag_begin + two));
+			if (parent_node != nullptr && parent_node->tag == close_tag)
+			{
+				current_node = parent_node;
+				parent_node = current_node->parent;
+				continue;
+			}
+		}
+		tag_close = line.find(symbol->end_tag_symbols, tag_end);
+		if (tag_close != std::string::npos)
+		{
+			element_data = line.substr(tag_end + one, tag_close - tag_end - one);
+		}
+		CreateElement(current_node, element_tag, element_data);
+		if (first_flag)
+		{
+			m_root = current_node;
+			parent_node = current_node;
+			first_flag = false;
+		}
+		else
+		{
+			ChildAppend(parent_node, current_node);
+			if (tag_close == std::string::npos)
+			{
+				parent_node = current_node;
+				current_node = nullptr;
+			}
+		}
+	}
+	file.close();
+	return true;
+}
+
+bool XMLParser::CreateElement(std::shared_ptr<XMLNode>& elem_pointer, const std::string& elem_tag, const std::string& elem_data)
+{
+	if (elem_tag.empty())
+	{
+		return false;
+	}
+	elem_pointer.reset(new XMLNode);
+	elem_pointer->tag = elem_tag;
+	elem_pointer->value = elem_data;
+	return true;
+}
+
+bool XMLParser::ChildAppend(std::shared_ptr<XMLNode>& parent, std::shared_ptr<XMLNode>& child)
+{
+	if (parent == nullptr || child == nullptr)
+	{
+		return false;
+	}
+	parent->children.push_back(child);
+	child->parent = parent;
+	return true;
+}
+
+bool XMLParser::FileCheck(const std::filesystem::path& filename) const
+{
+	if (!std::filesystem::exists(filename))
+	{
+		//std::cout << "File does not exist.\n";
+		return false;
+	}
+	if (!std::filesystem::is_regular_file(filename))
+	{
+		//std::cout << "File is not regular file.\n";
+		return false;
+	}
+	const std::string extension = ".xml";
+	if (std::filesystem::path(filename).extension().string() != extension)
+	{
+		//std::cout << "It's not XML file.\n";
+		return false;
+	}
+	if (std::filesystem::is_empty(filename))
+	{
+		//std::cout << "File is empty.\n";
+		return false;
+	}
+	return true;
+}
+
+void XMLParser::FindByTag(const std::shared_ptr<XMLNode>& current_node, const std::string& tag_needed, std::string& data)
+{
+	if (!current_node->children.size())
+	{
+		return;
+	}
+	for (size_t i = 0; i < current_node->children.size(); i++)
+	{
+
+		if (current_node->children[i]->tag == tag_needed)
+		{
+			data = current_node->children[i]->value;
+			break;
+		}
+		else
+		{
+			FindByTag(current_node->children[i], tag_needed, data);
+		}
+	}
+}
+
+bool XMLParser::ValueCheck(const std::string& line, unsigned int& number)
+{
+	std::string data = "";
+	FindByTag(m_root, line, data);
+	if (!data.empty())
+	{
+		try
+		{
+			number = stoi(data);
+		}
+		catch (...)
+		{
+			return false;
+		}
+		return true;
+	}
+	return false;
+}
+
+std::string XMLParser::GetServerName()
+{
+	if (m_root == nullptr)
+	{
+		ParseFile(XML_FILE_PATH);
+	}
+	std::string line = "servername";
+	std::string data = "";
```

Comment:
We don't need to initialize string value.

Reply:
Done.